mlir/CMakeLists.txt

```diff
@@ -121,6 +121,30 @@ FetchContent_Declare(tomlplusplus
 )
 FetchContent_MakeAvailable(tomlplusplus)
 
+################################################################################
+## Generate frontend_catalyst_version_py.h from frontend/catalyst/_version.py ##
+################################################################################
+
+function(make_version_header_file input_file output_file)
+    set(version_number "unknown")
+    if (NOT EXISTS ${input_file})
+        message(WARNING "Could not find ${input_file}")
+    else()
+        file(READ ${input_file} input_content)
+        if (NOT input_content MATCHES "__version__ = \"([^\"]+)\"")
+            message(WARNING "Could not find Catalyst version in ${input_file}")
+        else()
+            set(version_number "${CMAKE_MATCH_1}")
+        endif()
+    endif()
+    set(header_content "static constexpr const char *CATALYST_VERSION = \"${version_number}\";")
+    file(WRITE ${output_file} "${header_content}")
```

Comment:
I think this solves a similar problem to the build time macro we have configured here:
https://github.com/PennyLaneAI/catalyst/blob/f314679ccc577c425b45e8a3804beac93ba317c9/runtime/lib/capi/RuntimeCAPI.cpp#L142
https://github.com/PennyLaneAI/catalyst/blob/f314679ccc577c425b45e8a3804beac93ba317c9/runtime/lib/registry/CMakeLists.txt#L31

I don't know how the two compare though, so I'm happy to go with this one 👍